### PR TITLE
Adds Closed Sets

### PR DESCRIPTION
**Pull request by mronian**

### CODE REVIEW

**Comment by mronian** [comment]
cc @timholy , @ararslan :)

File: src/closed.jl [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	left, right = promote(left, right)
+	ClosedSet(left, right)
+end
+
+function union(A::ClosedSet, B::ClosedSet)
+	left = min(A.left, B.left)
+	right = max(A.right, B.right)
+	left, right = promote(left, right)
+	ClosedSet(left, right)
+end
```
Comment (mronian) [inline src/closed.jl]: Should `union` and `intersect` if any of the sets is empty?

File: src/closed.jl [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	left::T
+	right::T
+end
+
+function ClosedSet(left, right)
+	(I_left, I_right) = promote(left, right)
+	ClosedSet(I_left, I_right)
+end
+
+show(io::IO, I::ClosedSet) = print(io, "[", I.left, ",", I.right, "]")
```
Comment (mronian) [inline src/closed.jl]: How to test IO?

File: src/closed.jl [hunk 3]
```diff
@@ -0,0 +1,44 @@
+"""
+Defines the ClosedSet type.
+
+A ClosedSet(left, right) is an interval set where both bounds are closed i.e. the range is [left, right]
+"""
+immutable ClosedSet{T}
+	left::T
```
Comment (timholy) [inline src/closed.jl]: indentation 4 chars

File: src/closed.jl [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+end
+
+function ClosedSet(left, right)
+	(I_left, I_right) = promote(left, right)
+	ClosedSet(I_left, I_right)
+end
+
+show(io::IO, I::ClosedSet) = print(io, "[", I.left, ",", I.right, "]")
+
+length(I::ClosedSet) = I.right - I.left + 1
```
Comment (timholy) [inline src/closed.jl]: Doesn't work for non-integers. Maybe we shouldn't define `length` at all.

File: src/closed.jl [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+in(v, I::ClosedSet) = I.left <= v <= I.right
+
+isempty(A::ClosedSet) = A.left > A.right
+
+isequal(A::ClosedSet, B::ClosedSet) = (A.left == B.left && A.right == B.right) || isempty(A) && isempty(B)
```
Comment (timholy) [inline src/closed.jl]: Would be worth benchmarking whether this version or the non-shortcircuiting version (`&` instead of `&&` and `|` instead of `||`) is faster.

File: src/closed.jl [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+==(A::ClosedSet, B::ClosedSet) = isequal(A, B)
```
Comment (timholy) [inline src/closed.jl]: Do you need to define this? Why is `==` defined in terms of `isequal`?

File: src/closed.jl [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+function intersect(A::ClosedSet, B::ClosedSet)
+	left = max(A.left, B.left)
+	right = min(A.right, B.right)
+	left, right = promote(left, right)
```
Comment (timholy) [inline src/closed.jl]: You don't need this line, because the constructor will take care of that. Moreover, `left` and `right` should already be of the same type as each other, even if they differ from the type of `A` and/or `B`, because `min` and `max` promote.

File: test/runtests.jl [hunk 8]
```diff
@@ -1,5 +1,39 @@
-using IntervalSets
-using Base.Test
+module IntervalSetsTest
 
-# write your own tests here
-@test 1 == 2
+using IntervalSets, Base.Test, FactCheck
```
Comment (timholy) [inline test/runtests.jl]: Would it be OK to switch to Base.Test and `@testset`?

File: src/closed.jl (hunk 1 above)
Comment (timholy) [inline src/closed.jl]: `intersect` doesn't have to, just add some tests for intersection of empty sets and I bet you'll see it already works as-is.

More importantly, `union` needs to make sure that the intervals intersect; I think this implementation will "fill in" gaps between two disjoint intervals. It should throw an error (`InexactError`? `ArgumentError`?) when the two inputs don't intersect, because the union can't be expressed as a single closed set.

File: src/closed.jl (hunk 2 above)
Comment (timholy) [inline src/closed.jl]: See `IOBuffer` (and many examples in Julia's own `test/` folder).

File: REQUIRE [hunk 9]
```diff
@@ -1 +1 @@
-julia 0.4
+julia 0.5
```
Comment (timholy) [inline REQUIRE]: At present I think you need 0.5-, since 0.5 isn't released.

File: .travis.yml [hunk 10]
```diff
@@ -1,19 +1,16 @@
...
 language: julia
 os:
-  - linux
-  - osx
+    - linux
+    - osx
 julia:
-  - release
-  - nightly
+    - nightly
```
Comment (timholy) [inline .travis.yml]: Better make this
```
- 0.5
- nightly
```
because those two are already different from one another (and the differences will keep growing).

File: .travis.yml [hunk 11]
```diff
@@ -1,19 +1,16 @@
...
 notifications:
-  email: false
-# uncomment the following lines to override the default test script
-#script:
-#  - if [[ -a .git/shallow ]]; then git fetch --unshallow; fi
-#  - julia -e 'Pkg.clone(pwd()); Pkg.build("IntervalSets"); Pkg.test("IntervalSets"; coverage=true)'
+    email: false
+script:
+    - if [[ -a .git/shallow ]]; then git fetch --unshallow; fi
+    - julia -e 'Pkg.clone(pwd()); Pkg.build("IntervalSets")'
```
Comment (timholy) [inline .travis.yml]: Any reason to uncomment the original? I don't think this does anything different.

File: src/closed.jl [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+end
+
+function union(A::ClosedSet, B::ClosedSet)
+	left = min(A.left, B.left)
+	right = max(A.right, B.right)
+	left, right = promote(left, right)
+	ClosedSet(left, right)
+end
+
+issubset(A::ClosedSet, B::ClosedSet) =  isempty(A) || (A.left in B && A.right in B)
```
Comment (timholy) [inline src/closed.jl]: Likewise here about the short-circuiting benchmarking.

File: src/closed.jl (hunk 3 above)
Comment (ararslan) [inline src/closed.jl]: More specifically, 4 spaces. Indenting with 4 tabs would be chaos. 😄

File: .travis.yml [hunk 13]
```diff
@@ -1,19 +1,16 @@
-# Documentation: http://docs.travis-ci.com/user/languages/julia/
 language: julia
 os:
-  - linux
-  - osx
+    - linux
```
Comment (ararslan) [inline .travis.yml]: It doesn't really matter but out of curiosity why change the indentation style in the default Travis file?

Comment (mronian) [inline .travis.yml]: I copied the entire contents from a travis file in another package which probably lead to this :sweat_smile: I'll change it :smile:

File: .travis.yml (hunk 11 above)
Comment (mronian) [inline .travis.yml]: I copied the entire contents from a travis file in another package :sweat_smile:, which lead to these changes. Sorry about that! Will restore them :smile:

File: .travis.yml (hunk 13 above)
Comment (timholy) [inline .travis.yml]: Just checking, you know about `PkgDev.generate`, right?

Comment (mronian) [inline .travis.yml]: Yep. I'll clean this mess :smiley:

File: src/closed.jl (hunk 6 above)
Comment (mronian) [inline src/closed.jl]: Should `==` be only `(A.left == B.left && A.right == B.right)`  and not the `isempty` part of `isequal`?

File: src/closed.jl (hunk 12 above)
Comment (mronian) [inline src/closed.jl]: This becomes faster if `|` and `&` are used. (First benchmark is for the original one with `||` and `&&`)- 

```julia
julia> @benchmark issubset(A, B)
julia> @benchmark issubset(A, B)
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     995
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     27.00 ns (0.00% GC)
  median time:      27.00 ns (0.00% GC)
  mean time:        27.70 ns (0.00% GC)
  maximum time:     71.00 ns (0.00% GC)

BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     997
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     22.00 ns (0.00% GC)
  median time:      22.00 ns (0.00% GC)
  mean time:        22.40 ns (0.00% GC)
  maximum time:     49.00 ns (0.00% GC)
```

File: src/closed.jl (hunk 5 above)
Comment (mronian) [inline src/closed.jl]: This becomes slower with the non-shortcircuiting version. (First benchmark is for the one with `||` and `&&`) - 

```julia
julia> @benchmark isequal(A, B)
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     998
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     16.00 ns (0.00% GC)
  median time:      16.00 ns (0.00% GC)
  mean time:        16.49 ns (0.00% GC)
  maximum time:     43.00 ns (0.00% GC)
julia> @benchmark isequal(A, B)
BenchmarkTools.Trial: 
  samples:          10000
  evals/sample:     997
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  0.00 bytes
  allocs estimate:  0
  minimum time:     21.00 ns (0.00% GC)
  median time:      21.00 ns (0.00% GC)
  mean time:        21.91 ns (0.00% GC)
  maximum time:     66.00 ns (0.00% GC)
```

File: src/closed.jl (hunk 6 above)
Comment (timholy) [inline src/closed.jl]: Hmmm, interesting. Not entirely sure, but I wonder if `0.0 == -0.0` is justification for saying that all empty sets should return `true` for `==`.

It seems kind of odd that `==` is more permissive than `isequal` about the sign of 0 but less permissive about `NaN`.

File: src/closed.jl (hunk 12 above)
Comment (timholy) [inline src/closed.jl]: Better to use `$`, i.e., `@benchmark issubset($A, $B)`.

File: src/closed.jl (hunk 2 above)
Comment (timholy) [inline src/closed.jl]: Oh, and this should probably print as `left..right` since that's how they will be constructed (that way you can copy/paste output into the REPL and have it work).

File: src/closed.jl (hunk 1 above)
Comment (timholy) [inline src/closed.jl]: Case of floats is interesting because you should check against `nextfloat` (when the edges are touching but not overlapping).

File: src/closed.jl (hunk 6 above)
Comment (mronian) [inline src/closed.jl]: Shall I remove the `==` function?

File: src/closed.jl [hunk 14]
```diff
@@ -0,0 +1,54 @@
+"""
+Defines the ClosedSet type.
+
+A ClosedSet(left, right) is an interval set where both bounds are closed i.e. the range is [left, right]
```
Comment (ararslan) [inline src/closed.jl]: "Defines the ClosedSet type" is probably unnecessary since it should (hopefully) be obvious to a user that the type constructor constructs the type. :smile: 

I'd recommend putting backticks around `ClosedSet(left, right)` and perhaps rephrasing a bit as: "A `ClosedSet(left, right)` is an interval set that includes both its upper and lower bounds. In mathematical notation, the constructed range is `[left, right]`."

I say "in mathematical notation" there so as to avoid confusion with the syntax for arrays.

File: src/closed.jl [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+==(A::ClosedSet, B::ClosedSet) = isequal(A, B)
+
+function intersect(A::ClosedSet, B::ClosedSet)
+    left = max(A.left, B.left)
+    right = min(A.right, B.right)
+    ClosedSet(left, right)
+end
+
+function union{T<:AbstractFloat}(A::ClosedSet{T}, B::ClosedSet{T})
+    @assert max(A.left, B.left) <= nextfloat(min(A.right, B.right)) throw(ArgumentError("The sets are disjoint."))
```
Comment (ararslan) [inline src/closed.jl]: Rather than using `@assert`, I'd recommend the `condition || throw(...)` syntax. I'd also maybe make the error more specific, maybe something like `Cannot construct the union of disjoint sets`?

File: src/closed.jl (hunk 6 above)
Comment (timholy) [inline src/closed.jl]: No, on balance I think we should test equality, it just should be
```
==(A::ClosedSet, B::ClosedSet) = (A.left == B.left & A.right == B.right) | (isempty(A) & isempty(B))
```
or possibly the short-circuiting version. (I doubt it, though.)

File: src/closed.jl [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+Defines the ClosedSet type.
+
+A ClosedSet(left, right) is an interval set where both bounds are closed i.e. the range is [left, right]
+"""
+immutable ClosedSet{T}
+    left::T
+    right::T
+end
+
+function ClosedSet(left, right)
```
Comment (timholy) [inline src/closed.jl]: This can be a one-liner, if you wish. `@code_llvm` suggests that the 2-argument case does not suffer from splatting penalty (perhaps it's special-cased in the compiler).

File: src/closed.jl (hunk 6 above)
Comment (timholy) [inline src/closed.jl]: Oh, I see now, you have both `isequal` and `==`. It's hard to say which of these we should define, but I lean towards `==`. Or maybe both with `isequal` using `isequal` for comparing `left` and `right`, and `==` using `==` for the same comparisons.

**Comment by timholy** [comment]
Looking quite good (obviously we'll need some more documentation at some point). Nice work!

File: src/closed.jl (hunk 14 above)
Comment (simonbyrne) [inline src/closed.jl]: Also, why `ClosedSet` and not `ClosedInterval`?

Comment (ararslan) [inline src/closed.jl]: Simon makes a very good point. Formally speaking, a "closed set" is an *arbitrary* set that contains its limit points whereas a closed interval is much more specific (and is what is implemented).

Comment (timholy) [inline src/closed.jl]: I guess there's a slight conflict with https://github.com/scheinerman/ClosedIntervals.jl/blob/master/src/ClosedIntervals.jl, though. Still, it's undeniably the better term, so I guess we should use module names to disambiguate.

Comment (ararslan) [inline src/closed.jl]: I'm not super worried about the conflict with ClosedIntervals.jl since it isn't METADATA-registered (and would take some work to get ready for registration) nor does it appear to be all that actively developed.

Comment (mronian) [inline src/closed.jl]: I had the same concern as @timholy mentioned. But I guess the consensus is to use `ClosedIntervals` .

**Comment by mronian** [comment]
Did some benchmarks and it turns out that for `==`, the version with `&, |` is faster. For other functions it is the other way round. I have kept whichever is faster.

File: src/closed.jl [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+
+show(io::IO, I::ClosedInterval) = print(io, I.left, "..", I.right)
+
+in(v, I::ClosedInterval) = I.left <= v <= I.right
+
+isempty(A::ClosedInterval) = A.left > A.right
+
+isequal(A::ClosedInterval, B::ClosedInterval) = (isequal(A.left, B.left) & isequal(A.right, B.right)) | isempty(A) & isempty(B)
+
+==(A::ClosedInterval, B::ClosedInterval) = (A.left == B.left && A.right == B.right) || isempty(A) && isempty(B)
```
Comment (mronian) [inline src/closed.jl]: Is this fine?

**Comment by mronian** [comment]
What is an appropriate time to register a package? (I want to use this in other packages, but considering it has only `ClosedInterval` right now, will it be accepted?)

**Comment by ararslan** [comment]
In terms of registration, I think as long as some basic set of functionality is implemented (and that's reflected in the version number, e.g. 0.0.1) then whenever things are working is a fine time.

File: src/closed.jl (hunk 17 above)
Comment (timholy) [inline src/closed.jl]: Just checking, does the operator precedence work out? Might want to add a test `I == I` and/or `L == L` below.

**Comment by timholy** [comment]
Agreed, I don't think there's a minimum unit of functionality required for a package.
